File: api/models/chat_message.py
from django.db import models
from django.contrib.auth.models import User
from api.models.appointment import Appointment
from api.models.appointment_chat import AppointmentChat
from api.models.diagnosis import Diagnosis

class ChatMessage(models.Model):
    """
    Model representing a chat message.
    """

    sender = models.ForeignKey(User, on_delete=models.CASCADE)
    text = models.TextField(max_length=1000)
    chat = models.ForeignKey(AppointmentChat, on_delete=models.CASCADE)
    diagnosis = models.ForeignKey(Diagnosis, on_delete=models.SET_NULL, null=True)
    appointment = models.ForeignKey(Appointment, on_delete=models.SET_NULL, null=True)
    time = models.DateTimeField(auto_now_add=True)
    seen = models.BooleanField(default=False)
    
    def __str__(self):
        """
        Returns the string representation of the chat message.
        """
        return str(self.id) + ' ' + str(self.sender) + ' ' + self.text + ' ' + str(self.chat) + ' ' + str(self.diagnosis) + ' ' + str(self.appointment) + ' ' + str(self.time) + ' ' + str(self.seen)

File: khungulanga_app/lib/models/chat_message.dart
import 'appointment_chat.dart';
import 'diagnosis.dart';
import 'appointment.dart';
import 'user.dart';

class ChatMessage {
  final int? id;
  final User sender;
  final String text;
  final AppointmentChat chat;
  final Diagnosis? diagnosis;
  final Appointment? appointment;
  final DateTime time;
  final bool seen;

  ChatMessage({
    this.id,
    required this.sender,
    required this.text,
    required this.chat,
    this.diagnosis,
    this.appointment,
    required this.time,
    required this.seen,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['id'],
      sender: User.fromJson(json['sender']),
      text: json['text'],
      chat: AppointmentChat.fromJson(json['chat']),
      diagnosis: json['diagnosis'] != null ? Diagnosis.fromJson(json['diagnosis']) : null,
      appointment: json['appointment'] != null ? Appointment.fromJson(json['appointment']) : null,
      time: DateTime.parse(json['time']),
      seen: json['seen'],
    );
  }

  Map<String, dynamic> toJsonMap() {
    return {
      'id': id,
      'sender': sender.toJson(),
      'text': text,
      'chat': chat.id,
      'diagnosis': diagnosis?.id,
      'appointment': appointment?.id,
      'time': time.toIso8601String(),
      'seen': seen,
    };
  }
}